### PR TITLE
Update hamiltonians.jl

### PR DESCRIPTION
**Pull request by marwahaha**
I made a number of comments to hamiltonians.jl .

My main comments are:
* I think you should have a few more functions (if you find yourself saying "it does this if ___ otherwise ___" then there should be two functions)
* have a consistent way to write what each function returns (it's hard to parse from the explanatory text on top)
*  explain what your temp variables mean

### CODE REVIEW

File: src/hamiltonians.jl [hunk 1]
```diff
@@ -65,20 +74,26 @@ end
...
+        displacement::String,
         time_dependent_eta::Bool
     )
     b, indxs, cindxs = _setup_base_hamiltonian(
             T, timescale, lamb_dicke_order, rwa_cutoff, displacement, time_dependent_eta
         )
     aui, gbi, gbs, bfunc, δνi, δνfuncs = _setup_fluctuation_hamiltonian(T, timescale)
     S = SparseOperator(get_basis(T))
-    function f(t, ψ)  # a two argument function is required in the QuantumOptics solvers
+    # we return the function f (2 arguments required for QuantumOptics solvers)
```
Comment (marwahaha) [inline src/hamiltonians.jl:93]: question

File: src/hamiltonians.jl [hunk 2]
```diff
@@ -65,20 +74,26 @@ end
...
+    function f(t, ψ)
         @inbounds begin
+            #? Is indxs a rectangle? We could precompute length(indxs) and length(indxs[i])
```
Comment (marwahaha) [inline src/hamiltonians.jl:96]: question

File: src/hamiltonians.jl [hunk 3]
```diff
@@ -159,6 +176,7 @@ additional vector of vectors of indices.
 
 Finally, since we require the Hamiltonian to be Hermitian, h[i, j] = conj(h[j, i]), this 
 function does not keeps track of only one of these pairs.
+#? Does it keep track of both or one?
```
Comment (marwahaha) [inline src/hamiltonians.jl:179]: question

File: src/hamiltonians.jl [hunk 4]
```diff
@@ -175,7 +193,9 @@ function _setup_base_hamiltonian(
     
     ηm, Δm, Ωm = _ηmatrix(T), _Δmatrix(T, timescale), _Ωmatrix(T, timescale)
     lamb_dicke_order = _check_lamb_dicke_order(lamb_dicke_order, L)
+    #? what is a ld_array?
```
Comment (marwahaha) [inline src/hamiltonians.jl:196]: question

File: src/hamiltonians.jl [hunk 5]
```diff
@@ -175,7 +193,9 @@ function _setup_base_hamiltonian(
...
     ld_array, rows, vals = _ld_array(mode_dims, lamb_dicke_order, νlist, timescale)
+    #? what does the below code do?
```
Comment (marwahaha) [inline src/hamiltonians.jl:198]: question

File: src/hamiltonians.jl [hunk 6]
```diff
@@ -192,6 +212,7 @@ function _setup_base_hamiltonian(
         if m ≡ 1
             rn = length(ions) - n + 1
             ts = _transitions(ions[n])
+            #? what is C here?
```
Comment (marwahaha) [inline src/hamiltonians.jl:215]: question

File: src/hamiltonians.jl [hunk 7]
```diff
@@ -203,6 +224,7 @@ function _setup_base_hamiltonian(
             ion_reps = Int(Q / size(C, 1))
         end
         ηnm = view(ηm, rn, m, :) 
+        #? what is this for? to decide whether it's a vector or single number?
```
Comment (marwahaha) [inline src/hamiltonians.jl:227]: question

File: src/hamiltonians.jl [hunk 8]
```diff
@@ -257,6 +281,7 @@ function _setup_base_hamiltonian(
                     pushfirst!(s_cri, (-1 * isodd(parity), 0))
                 end
                 
+                #? I would use layered if statements, if (time dependent eta), if (truncated) do ___ else do ____ , else ___
```
Comment (marwahaha) [inline src/hamiltonians.jl:284]: suggestion

File: src/hamiltonians.jl [hunk 9]
```diff
@@ -221,12 +243,14 @@ function _setup_base_hamiltonian(
         for (ti, tr) in enumerate(ts)
             Δ, Ω = Δm[rn, m][ti], Ωm[rn, m][ti]
             Δ_2π = Δ/2π
+            #? I haven't seen this syntax, what does it do?
```
Comment (marwahaha) [inline src/hamiltonians.jl:246]: question

File: src/hamiltonians.jl [hunk 10]
```diff
@@ -221,12 +243,14 @@ function _setup_base_hamiltonian(
...
             typeof(Ω) <: Number && continue  # e.g. the laser doesn't shine on this ion
             locs = view(ion_idxs, (ti - 1) * ion_reps + 1 : ti * ion_reps)
             for j in 1:prod(mode_dims); for i in nzrange(ld_array, j)
                 ri = rows[i]
                 ri<j && continue
                 cf = vals[i]
+                # I think the below rwa could be rewritten, pflag and nflag seem like weird variable names.
```
Comment (marwahaha) [inline src/hamiltonians.jl:253]: suggestion

File: src/hamiltonians.jl [hunk 11]
```diff
@@ -268,6 +293,7 @@ function _setup_base_hamiltonian(
                 row, col = s_ri[1]
                 if haskey(indxs_dict, s_ri[1]) 
                     # this will happen when multiple lasers address the same transition
+                    #? why would this happen? could this be defined outside of this method?
```
Comment (marwahaha) [inline src/hamiltonians.jl:296]: question

File: src/hamiltonians.jl [hunk 12]
```diff
@@ -383,6 +409,7 @@ end
 # _setup_δν_hamiltonian for use in the hamiltonian function. The one additional task 
 # performed is the creation of an array of indices (all_unique_indices), which keeps track
 # of all the diagonal indices affected by δν and/or δB, which is useful in hamiltonian().
+#? we should label each one of these, and explain what they are for (gbi, gbs, etc have names slightly too short for me)
```
Comment (marwahaha) [inline src/hamiltonians.jl:412]: suggestion

File: src/hamiltonians.jl [hunk 13]
```diff
@@ -561,6 +590,7 @@ function _inv_get_kron_indxs(indxs, dims)
 end
 
 # similar to _Dnm, but meant to be used when η is assumed constant in ξ=iηe^(i2πνt)
+#? can we place this function near _Dnm?
```
Comment (marwahaha) [inline src/hamiltonians.jl:593]: suggestion

File: src/hamiltonians.jl [hunk 14]
```diff
@@ -576,6 +606,7 @@ end
 # If lamb_dicke_order is <: Int, this constructs a constant vector with this value of length
 # L (i.e. same lamb_dicke_order for all modes). Otherwise lamb_dicke_order is reversed and 
 # returned.
+#? why do we reverse the lamb dicke order? That seems really confusing.
```
Comment (marwahaha) [inline src/hamiltonians.jl:609]: suggestion

**Comment by jbroz11** [comment]
We will be changing this code a bit when we add in arb rotating frame transformation capability, so this is no longer relevant.
